cachi2/core/package_managers/yarn_classic/utils.py

```diff
@@ -0,0 +1,101 @@
+from collections import deque
+from typing import Any
+
+from pyarn.lockfile import Package as PYarnPackage
+
+from cachi2.core.package_managers.yarn_classic.project import PackageJson, YarnLock
+from cachi2.core.package_managers.yarn_classic.workspaces import Workspace
+from cachi2.core.rooted_path import RootedPath
+
+
+def find_non_dev_deps(
+    main_package_json: PackageJson,
+    yarn_lock: YarnLock,
+    workspaces: list[Workspace],
+) -> set[str]:
+    """
+    Identify all non-dev dependencies in the root package and its workspaces.
+
+    A dependency is classified as "non-dev" if:
+    - It is listed in `dependencies`, `peerDependencies`, or `optionalDependencies`
+      of any `package.json` file.
+    - It is a transitive dependency of another non-dev dependency.
+
+    A dependency is classified as "dev" if:
+    - It is listed in `devDependencies` of any `package.json` file.
+    - It is a transitive dependency of a dev dependency.
+
+    This function ensures that a dependency appearing as both a dev and non-dev
+    dependency is marked as non-dev.
+    """
+    # TODO: PackageJson object should be an attribute of the Workspace object
+    ws_package_jsons = [
+        PackageJson.from_file(RootedPath(ws.path / "package.json")) for ws in workspaces
```

Comment:
I know, but I think it would be much clearer to use `PackageJson` instead. At the time when workspaces were implemented there was no `PackageJson` class.

Reply:
You are right, turns out Workspaces predate PackageJson.